api/src/main/java/com/epam/pipeline/manager/git/github/GitHubService.java

```diff
@@ -0,0 +1,412 @@
+                .map(mapper::tagToRevision)
+                .orElse(null);
+    }
+
+    @Override
+    public Revision getLastRevision(final Pipeline pipeline, final String ref) {
+        final Response<List<GitHubCommitNode>> commits = getClient(pipeline).getLastCommit();
+        return Optional.ofNullable(commits.body()).orElse(Collections.emptyList()).stream()
+                .findFirst()
+                .map(mapper::commitToRevision)
+                .orElse(null);
+    }
+
+    @Override
+    public GitCredentials getCloneCredentials(final Pipeline pipeline, final boolean useEnvVars,
+                                              final boolean issueToken, final Long duration) {
+        final GitRepositoryUrl repositoryUrl = GitRepositoryUrl.from(pipeline.getRepository());
+        final String token = pipeline.getRepositoryToken();
+        final String username = preferenceManager.getPreference(SystemPreferences.BITBUCKET_CLOUD_USER_NAME);
+        final String host = repositoryUrl.getHost();
+        return GitCredentials.builder()
+                .url(GitRepositoryUrl.asString(repositoryUrl.getProtocol(), username, token, host,
+                        repositoryUrl.getNamespace().orElseThrow(() -> buildUrlParseError(PROJECT_NAME)),
+                        repositoryUrl.getProject().orElseThrow(() -> buildUrlParseError(REPOSITORY_NAME))))
+                .userName(username)
+                .token(token)
+                .build();
+    }
+
+    @Override
+    public GitCommitEntry getCommit(final Pipeline pipeline, final String commitId) {
+        final GitHubCommitNode commit = getClient(pipeline).getCommit(commitId);
+        return Optional.ofNullable(commit)
+                .map(mapper::commitToCommitEntry)
+                .orElse(null);
+    }
+
+    @Override
+    public GitTagEntry getTag(final Pipeline pipeline, final String tagName) {
+        final GitHubRelease tag = getClient(pipeline).getTag(tagName);
+        return Optional.ofNullable(tag)
+                .map(mapper::tagToTagEntry)
+                .orElse(null);
+    }
+
+    @Override
+    public List<GitRepositoryEntry> getRepositoryContents(final Pipeline pipeline, final String rawPath,
+                                                          final String version, final boolean recursive) {
+        final GitHubClient client = getClient(pipeline);
+
+        final String path = ProviderUtils.DELIMITER.equals(rawPath) ? version : getContentSha(rawPath, version, client);
+
+        final GitHubTree tree = client.getTree(path, !recursive ? null : true);
+        final List<GitHubTreeContent> values = tree.getTree();
+        final Stream<GitHubTreeContent> folders = values.stream()
+                .filter(v -> v.getType().equals(GITHUB_FOLDER_MARKER));
+
+        final Stream<GitHubTreeContent> files = values.stream()
+                .filter(v -> v.getType().equals(GITHUB_FILE_MARKER));
+
+        return Stream.concat(folders, files)
```

Comment:
Do we need separate streams here? Can we use `filter` instead?
```suggestion
        return values.stream()
                .filter(v -> v.getType().equals(GITHUB_FOLDER_MARKER) || v.getType().equals(GITHUB_FILE_MARKER))
```

Reply:
The separation is because we have to put folders in the beginning of the content list